OpenEdXMobile/src/main/java/org/edx/mobile/view/BulkDownloadFragment.java

```diff
@@ -0,0 +1,428 @@
+    private Timer downloadProgressTimer;
+    private Timer deleteVideosTimer;
+
+    /**
+     * Summarises the download status of all the videos within a course.
+     */
+    private CourseVideosStatus videosStatus = new CourseVideosStatus();
+    /**
+     * List of videos that can be downloaded.
+     */
+    private List<HasDownloadEntry> remainingVideos = new ArrayList<>();
+    /**
+     * List of videos that are currently being downloaded or haven been downloaded.
+     */
+    private List<VideoModel> removableVideos = new ArrayList<>();
+
+    public BulkDownloadFragment() {
+    }
+
+    @SuppressLint("ValidFragment")
+    public BulkDownloadFragment(@NonNull NewCourseOutlineAdapter.DownloadListener downloadListener,
+                                @NonNull IEdxEnvironment environment) {
+        this.downloadListener = downloadListener;
+        this.environment = environment;
+    }
+
+    @Override
+    public View onCreateView(LayoutInflater inflater, ViewGroup container,
+                             Bundle savedInstanceState) {
+        binding = DataBindingUtil.inflate(inflater, R.layout.row_bulk_download, container, false);
+        return binding.getRoot();
+    }
+
+    @Override
+    public void onViewCreated(View view, Bundle savedInstanceState) {
+        super.onViewCreated(view, savedInstanceState);
+        prefManager = new VideoPrefs(getContext());
+
+        ViewCompat.setImportantForAccessibility(binding.getRoot(), ViewCompat.IMPORTANT_FOR_ACCESSIBILITY_NO);
+        ViewCompat.setImportantForAccessibility(binding.icon, ViewCompat.IMPORTANT_FOR_ACCESSIBILITY_NO);
+        ViewCompat.setImportantForAccessibility(binding.title, ViewCompat.IMPORTANT_FOR_ACCESSIBILITY_NO);
+        ViewCompat.setImportantForAccessibility(binding.description, ViewCompat.IMPORTANT_FOR_ACCESSIBILITY_NO);
+        ViewCompat.setImportantForAccessibility(binding.downloadProgress, ViewCompat.IMPORTANT_FOR_ACCESSIBILITY_NO);
+        ViewCompat.setImportantForAccessibility(binding.downloadButton, ViewCompat.IMPORTANT_FOR_ACCESSIBILITY_YES);
+        initDownloadSwitch();
+        if (courseComponent != null) {
+            populateViewHolder(courseComponent);
+        }
+    }
+
+    public void populateViewHolder(@NonNull CourseComponent courseComponent) {
+        this.courseComponent = courseComponent;
+        if (binding == null) return;
+        remainingVideos.clear();
+        removableVideos.clear();
+        videosStatus.reset();
+        videosStatus.courseId = courseComponent.getCourseId();
+
+        // Get all the videos of this course
+        final List<CourseComponent> allVideosInCourse = courseComponent.getVideos(true);
+        videosStatus.total = allVideosInCourse.size();
+
+        // Get all the videos of this course from DB that we have downloaded or are downloading
+        final List<VideoModel> videosInDB = environment.getDatabase().getAllVideosByCourse(courseComponent.getCourseId(), null);
```

Comment:
We should perform this db call asynchronously and show the spiral on Bulk download view list item till the time we prepare the data for the UI.

Reply:
Good point .. showing a spiral would require some extra effort, for now, I've hidden the view initially and show it when the db call returns.